src/log.rs

```diff
@@ -1,47 +1,152 @@
+    }
 }
 
-impl Default for Rav1dLogger {
+impl Default for Dav1dLogger {
     fn default() -> Self {
-        Self {
-            cookie: ptr::null_mut(),
-            callback: Some(rav1d_log_default_callback),
+        // Safety: `callback` is [`None`].
+        unsafe { Self::new(ptr::null_mut(), None) }
+    }
+}
+
+/// A [`Dav1dLogger`] from C that's not
+/// just a [`Rav1dLogger`] converted to a [`Dav1dLogger`].
+#[derive(Clone)]
+pub(crate) struct OnlyDav1dLogger {
+    cookie: *mut c_void,
+    callback: Dav1dLoggerCallback,
+}
+
+impl fmt::Write for OnlyDav1dLogger {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        // `s` doesn't have a terminating nul-byte,
+        // and it may have internal nul-bytes,
+        // so it's easiest just to print one byte at a time.
+        // This may be slow, but logging can be disabled if it's slow,
+        // or the Rust API can be used instead.
+        let fmt = cstr!("%c");
+        for &byte in s.as_bytes() {
+            // # Safety
+            //
+            // The first argument is `self.cookie`
+            // and the rest are safe to call `printf` with,
+            // as required by [`Self::new`].
+            unsafe {
+                (self.callback)(self.cookie, fmt.as_ptr(), byte as c_uint);
+            }
         }
+        Ok(())
     }
 }
 
-#[cold]
-pub unsafe extern "C" fn rav1d_log(c: *mut Rav1dContext, format: *const c_char, args: ...) {
-    if validate_input!(!c.is_null()).is_err() {
-        return;
+#[derive(Clone, Default)]
+pub(crate) enum Rav1dLogger {
+    Dav1d(OnlyDav1dLogger),
+    Stdout,
+    #[default]
+    Stderr,
+}
+
+/// Any type implementing [`Rav1dLog`] can be used with [`write!`].
+pub trait Rav1dLog {
+    fn write_fmt(&self, args: fmt::Arguments);
+}
+
+impl Rav1dLog for Rav1dLogger {
+    /// Logging doesn't have to be fast when it's on,
+    /// but we don't want it slow things down when it's off,
+    /// so ensure the logging code isn't inlined everywhere, bloating call sites.
+    #[inline(never)]
+    fn write_fmt(&self, args: fmt::Arguments) {
+        match self {
+            // The `dav1d.clone()` is because [`fmt::Write::write_fmt`] takes `&mut`
+            // even though we don't need it to.
+            // [`OnlyDav1dLogger`] is trivial to [`Clone`], though, so we can just do that.
+            Self::Dav1d(dav1d) => dav1d.clone().write_fmt(args).unwrap(),
+            Self::Stdout => stdout().write_fmt(args).unwrap(),
+            Self::Stderr => stderr().write_fmt(args).unwrap(),
+        }
     }
-    if ((*c).logger.callback).is_none() {
-        return;
+}
+
+impl Rav1dLog for Option<Rav1dLogger> {
+    /// When a logger isn't set, we don't want to have to run any code here,
+    /// so force this to be inlined so a [`None`] can be seen.
+    #[inline(always)]
+    fn write_fmt(&self, args: fmt::Arguments) {
+        if let Some(logger) = self {
+            logger.write_fmt(args);
+        }
+    }
```

Comment:
An alternative I think we should consider: We could replace the `rav1d_log` function with a `rav1d_log` macro that takes a `Rav1dContext` argument and includes the `if let Some(logger) = c.logger`. That way we don't need this extra trait or most of this infrastructure, as the logger can just be something that implements `fmt::Write` then.

Reply:
What advantage does that have?  That just seems more complex.  We can't implement `fmt::Write` due to the `&mut self` so we need our own function regardless, plus this way we don't return a `Result` that needs to be unwrapped and that bloats the code a lot for something that might never be run.  So at that point it's just a function vs. a macro and a function seems simpler.

Reply:
We don't need the trait at all, we could just have a method on Rav1dLogger that takes a `fmt::Argmuents` that is called inside the macro after using `format_args!` to construct the arguments. That's exactly what `writeln!` is doing, this would just be a specialized version of `writeln!`.

Thinking more on this, a macro would give us a bit more flexibility in the future if we want to change the structure of logging. I'm not set on it, but it seems more straightforward than the trait to me. This is also one place where I'm a bit concerned about code size and performance. We shouldn't spend time optimizing this right now though, but if we do decide we need to change things in the future, a macro gives us the flexibility to do so.

Reply:
I don't think it's worth it to reimplement `write!`, `writeln!`, etc.  Remember we'd have to implement the base and `*ln!` variants, and I think it's a lot clearer if we stick to the standard macros for this rather than inventing our own.  A trait isn't very complex and it comes with documentation on why the signature is why it is, which is also useful.